contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,171 @@
+
+3. Reusability
+As per DRY Principle, module allows developers to reuse their code that is contained in a module.
+
+## How Is It Implemented?
+
+Historically, JavaScript did not support modules up until 2015 with the inception of EcmaScript6 (ES6). Nevertheless, there have been "unofficial implementations" of module sytem in JavaScript in form of Module Pattern and CommonJS.
+
+This article will expand on those 3 implementations.
+
+### Module Pattern
+
+This is the way JavaScript developer create module in the past before the inception of CommonJS and ES6. Using a technique in JavaScript called IIFE (Immediately-Invoked Function Expression), they wrap their code in an IIFE.
+
+> IIFE syntax </br>
+> `(function() {})()`
+
+`anExampleModule.js`
+```javascript
+var anExampleModule = (function() {
+    var variableOne = 1;
+    var variableTwo = 2;
+
+    var getVariableOne = function() {
+        return variableOne;
+    }
+
+    var sumOfVariable = function() {
+        return variableOne + variableTwo;
+    }
+
+    var counter = function() {
+        variableOne++;
+    }
+
+    return {
+        getVariableOne: getVariableOne,
+        sumOfVariable: sumOfVariable,
+        counter,
+    }
+})()
+```
+
+`index.js`
+```javascript
+anExampleModule.getVariableOne(); // 1
+anExampleModule.sumOfVariable(); // 3
+```
+
+`index.html`
+```html
+<script src="./anExampleModule.js" />
+<script src="./index.js" />
+```
+
+In the above example, we are creating a module called `anExampleModule` in a separate `anExampleModule.js` file, which is used by `index.js` file.
+
+To include the module in `index.js`, we can just use the `<script>` tag in HTML to import it.
+
+### CommonJS
+
+Implemented by NodeJS, it is one of the most popular "unofficial" modules implementation. It uses the syntax `module.export` and `require`.
+
+`anExampleModule.js`
+```Javascript
+var variableOne = 1;
+var variableTwo = 2;
+
+getVariableOne = function() {
+  return variableOne;
+}
+
+sumOfVariable = function() {
+    return variableOne + variableTwo;
+}
+
+module.exports = {
+  getVariableOne: getVariableOne,
+  sumOfVariable: sumOfVariable,
+};
+```
+
+`index.js`
+```Javascript
+var anExampleModule = require('./anExampleModule.js');
+
+anExampleModule.getVariableOne(); // 1
+anExampleModule.sumOfVariable(); // 3
+```
+
+### EcmaScript2015 / ES6 modules
+
+At 2015, JavaScript introduced built-in modules in ES6. It uses the syntax `import` and `export`.
+
+`anExampleModule.js`
+```Javascript
+var variableOne = 1;
+var variableTwo = 2;
+
+export getVariableOne() {
+    return variableOne;
+}
+
+export sumOfVariable() {
+    return variableOne + variableTwo;
+}
+```
+
+`index.js`
+```Javascript
+import * as anExampleModule from './anExampleModule.js';
```

Comment:
just importing some stuff?

`import { getVariableOne } from './anExampleModule.js'`

Reply:
I think better to import all rather than import some stuff as that will involve object destructuring, another concept I think need to be explained? haha